README
NAME
    Linux::Smaps::Tiny - A minimal and fast alternative to Linux::Smaps

SYNOPSIS
        use Linux::Smaps::Tiny qw(get_smaps_summary);

        my $summary = get_smaps_summary();
        my $size = $summary->{Size};
        my $shared_clean = $summary->{Shared_Clean};
        my $shared_dirty = $summary->{Shared_Dirty};

        warn "Size / Clean / Dirty = $size / $shared_clean/ $shared_dirty";

DESCRIPTION
    This module is a tiny interface to /proc/PID/smaps files. It was written
    because when we rolled out Linux::Smaps in some critical code at a Big
    Internet Company we experienced slowdowns that were solved by writing a
    more minimal version.

    This module will try to use XS code to parse the smaps file, and if that
    doesn't work it'll fall back on a pure-Perl version.

    If something like that isn't your use case you should probably use
    Linux::Smaps instead. Also note that Linux::Smaps itself has been
    optimized
    <http://mail-archives.apache.org/mod_mbox/perl-modperl/201103.mbox/brows
    er> since this module was initially written.

  SPEED
    The distribution comes with a contrib/benchmark.pl script. As of writing
    this is the speed of Linux::Smaps v.s. Linux::Smaps::Tiny, both the XS
    and PP versions:

                                 Rate Linux::Smaps Linux::Smaps::Tiny::PP Linux::Smaps::Tiny
        Linux::Smaps            810/s           --                   -22%               -61%
        Linux::Smaps::Tiny::PP 1033/s          28%                     --               -51%
        Linux::Smaps::Tiny     2101/s         159%                   103%                 --

FUNCTIONS
  get_smaps_summary
    Takes an optional process id (defaults to "self") returns a summary of
    the smaps data for the given process. Dies if the process does not
    exist.

    Returns a hashref like this:

            {
              'MMUPageSize' => '184',
              'Private_Clean' => '976',
              'Swap' => '0',
              'KernelPageSize' => '184',
              'Pss' => '1755',
              'Private_Dirty' => '772',
              'Referenced' => '2492',
              'Size' => '5456',
              'Shared_Clean' => '744',
              'Shared_Dirty' => '0',
              'Rss' => '2492'
            };

    Values are in kB.

LICENSE AND COPYRIGHT
    Copyright 2011 Yves Orton <yves@cpan.org> and Ævar Arnfjörð Bjarmason
    <avar@cpan.org>

    This program is free software, you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Linux::Smaps::Tiny

0.09 2012-07-01 15:39:17

    - CPAN RT #78029: fix a memory leak reported by Norbert
      Buchmüller.

0.08 2011-07-20 19:20:57

    - The 0.07 releas had a bugfix in the XS API. It now actually
      works as documented when given an argument.

    - Minor documentation nits.

0.07 2011-07-20 19:14:45

    - Check errno right after fopen()

0.06 2011-04-03 19:18:44

    - We now parse smaps with XS code, and fall back to PP if we can't
      can_cc().

0.05 2011-03-30 10:27:33

    - Add a =cut before the 1; in Tiny.pm

0.04 2011-03-30 09:24:57

    - Update documentation

0.03 2011-03-29 12:52:03

    - Add more information to our die() messages

0.02 2011-03-29 12:18:55

    - Default to /proc/self/smaps, not /proc/$$/smaps

0.01 2011-03-29 09:16:11

    - Initial buggy version


Makefile.PL
# This Makefile.PL for  was generated by Dist::Zilla.
# Don't edit it but the dist.ini used to construct it.

use strict;
use warnings;
use ExtUtils::MakeMaker 6.30;

my %WriteMakefileArgs = (
  'ABSTRACT' => 'A minimal and fast alternative to L<Linux::Smaps>',
  'AUTHOR' => 'Ævar Arnfjörð Bjarmason <avar@cpan.org>',
  'BUILD_REQUIRES' => {
    'File::Find' => '0',
    'File::Temp' => '0',
    'List::Util' => '0',
    'Test::More' => '0'
  },
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::MakeMaker' => '6.30'
  },
  'DISTNAME' => 'Linux-Smaps-Tiny',
  'EXE_FILES' => [],
  'LICENSE' => 'perl',
  'NAME' => 'Linux::Smaps::Tiny',
  'PREREQ_PM' => {
    'Exporter' => '0',
    'XSLoader' => '0',
    'strict' => '0',
    'warnings' => '0'
  },
  'VERSION' => '0.09',
  'test' => {
    'TESTS' => 't/*.t'
  }
);

unless (can_cc()) {
    # Same parameters as http://cpansearch.perl.org/src/GBARR/Scalar-List-Utils-1.23/Makefile.PL
    $WriteMakefileArgs{XS}     = {};
    $WriteMakefileArgs{C}      = [];
    $WriteMakefileArgs{OBJECT} = '';
}


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);

# Copied from http://cpansearch.perl.org/src/GBARR/Scalar-List-Utils-1.23/Makefile.PL
sub can_cc {

    foreach my $cmd (split(/ /, $Config::Config{cc})) {
        my $_cmd = $cmd;
        return $_cmd if (-x $_cmd or $_cmd = MM->maybe_command($_cmd));

        for my $dir ((split /$Config::Config{path_sep}/, $ENV{PATH}), '.') {
            my $abs = File::Spec->catfile($dir, $_[1]);
            return $abs if (-x $abs or $abs = MM->maybe_command($abs));
        }
    }

    return;
}


lib/Linux/Smaps/Tiny.pm
package Linux::Smaps::Tiny;
BEGIN {
  $Linux::Smaps::Tiny::AUTHORITY = 'cpan:AVAR';
}
{
  $Linux::Smaps::Tiny::VERSION = '0.09';
}
use strict;
use warnings FATAL => "all";

BEGIN {
    local ($@, $!);
    eval {
        require XSLoader;
        XSLoader::load(__PACKAGE__, $Linux::Smaps::Tiny::VERSION || '0.01');
    };
}

use Exporter 'import';

our @EXPORT_OK = qw(get_smaps_summary);

=encoding utf8

=head1 NAME

Linux::Smaps::Tiny - A minimal and fast alternative to L<Linux::Smaps>

=head1 SYNOPSIS

    use Linux::Smaps::Tiny qw(get_smaps_summary);

    my $summary = get_smaps_summary();
    my $size = $summary->{Size};
    my $shared_clean = $summary->{Shared_Clean};
    my $shared_dirty = $summary->{Shared_Dirty};

    print "Size / Clean / Dirty = $size / $shared_clean / $shared_dirty\n";

=head1 DESCRIPTION

This module is a tiny interface to F</proc/PID/smaps> files. It was
written because when we rolled out L<Linux::Smaps> in some critical
code at a Big Internet Company we experienced slowdowns that were
solved by writing a more minimal version.

This module will try to use XS code to parse the smaps file, and if
that doesn't work it'll fall back on a pure-Perl version.

If something like that isn't your use case you should probably use
L<Linux::Smaps> instead. Also note that L<Linux::Smaps> itself L<has
been
optimized|http://mail-archives.apache.org/mod_mbox/perl-modperl/201103.mbox/browser>
since this module was initially written.

=head2 SPEED

The distribution comes with a F<contrib/benchmark.pl> script. As of
writing this is the speed of L<Linux::Smaps>
v.s. L<Linux::Smaps::Tiny>, both the XS and PP versions:

                             Rate Linux::Smaps Linux::Smaps::Tiny::PP Linux::Smaps::Tiny
    Linux::Smaps            810/s           --                   -22%               -61%
    Linux::Smaps::Tiny::PP 1033/s          28%                     --               -51%
    Linux::Smaps::Tiny     2101/s         159%                   103%                 --

=head1 FUNCTIONS

=head2 get_smaps_summary

Takes an optional process id (defaults to C<self>) returns a summary
of the smaps data for the given process. Dies if the process does not
exist.

Returns a hashref like this:

        {
          'MMUPageSize' => '184',
          'Private_Clean' => '976',
          'Swap' => '0',
          'KernelPageSize' => '184',
          'Pss' => '1755',
          'Private_Dirty' => '772',
          'Referenced' => '2492',
          'Size' => '5456',
          'Shared_Clean' => '744',
          'Shared_Dirty' => '0',
          'Rss' => '2492'
        };

Values are in kB.

=cut

unless (defined &get_smaps_summary) {
    require Linux::Smaps::Tiny::PP;
    *get_smaps_summary = \&Linux::Smaps::Tiny::PP::__get_smaps_summary;
}

=head1 LICENSE AND COPYRIGHT

Copyright 2011 Yves Orton <yves@cpan.org> and Ævar Arnfjörð Bjarmason
<avar@cpan.org>

This program is free software, you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

1;
